include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp

```diff
@@ -354,30 +350,46 @@ template<int dim, typename Number>
 void
 MultigridPreconditionerBase<dim, Number>::initialize_mapping()
 {
-  // We only need to initialize the mapping for all multigrid h-levels if it is of type
-  // dealii::MappingQCache (including MappingDoFVector as a derived class), while MappingQ is
-  // unproblematic.
-  std::shared_ptr<dealii::MappingQCache<dim> const> mapping_q_cache =
-    std::dynamic_pointer_cast<dealii::MappingQCache<dim> const>(mapping);
-
-  if(data.involves_h_transfer() and mapping_q_cache.get() != 0)
+  if(data.involves_h_transfer())
   {
-    if(multigrid_variant == MultigridVariant::GlobalCoarsening)
-    {
-      MappingTools::initialize_multigrid(coarse_grid_mappings,
-                                         mapping_q_cache,
-                                         coarse_triangulations);
-    }
-    else if(multigrid_variant == MultigridVariant::LocalSmoothing)
+    coarse_grid_mappings.resize(this->get_number_of_h_levels());
+
+    // We only need to explicitly initialize the mapping for all multigrid h-levels if it is of type
+    // dealii::MappingQCache (including MappingDoFVector as a derived class)
+    std::shared_ptr<dealii::MappingQCache<dim> const> mapping_q_cache =
+      std::dynamic_pointer_cast<dealii::MappingQCache<dim> const>(mapping);
+
+    if(mapping_q_cache.get() != 0)
     {
-      mapping_dof_vector =
-        std::make_shared<MappingDoFVector<dim, Number>>(mapping_q_cache->get_degree());
+      if(multigrid_variant == MultigridVariant::GlobalCoarsening)
+      {
+        MappingTools::initialize_multigrid<dim, Number>(coarse_grid_mappings,
+                                                        mapping_q_cache,
+                                                        coarse_triangulations);
+      }
+      else if(multigrid_variant == MultigridVariant::LocalSmoothing)
+      {
+        std::shared_ptr<MappingDoFVector<dim, Number>> mapping_dof_vector =
+          std::make_shared<MappingDoFVector<dim, Number>>(mapping_q_cache->get_degree());
+
+        MappingTools::initialize_multigrid(mapping_dof_vector, mapping_q_cache, *triangulation);
```

Comment:
Seeing this code, I'm wondering if we should let `initialize_multigrid` return `std::shared_ptr<MappingDoFVector<dim, Number>>` instead of passing it in as first argument (also for the other function). To me, it would look more intuitive if the computed quantity is returned, and in fact both cases seem to throw away the previous content anyway. This must not necessarily happen in this PR, though.

Reply:
Ah, that's again one of those cases. I believe it will not work automatically. We could of course return `std::vector<std::shared_ptr<dealii::Mapping>>` directly, but then it is maybe less clear how it gets computed.

Reply:
It makes sense in my opinion to unify the interfaces already in this PR. I talked to  @peterrum and we discussed to consider the case with a vector of mappings for coarse h-levels as the default case/interface and to consider dealii implementations with a single triangulation/mapping (for all h-levels) as legacy code supported internally.